### PR TITLE
Make most variants of module.exports = { ... } fixable

### PR DESCRIPTION
**Pull request by papandreou**
This is https://github.com/mysticatea/eslint-plugin-node/pull/208 against your fork now that `eslint-plugin-node` is unmaintained and `eslint-plugin-n` is being used by `eslint-config-standard` 😇 

It got reviewed by the original author, who then disappeared before merging it.

### CODE REVIEW

**Comment by papandreou** [comment]
CI failure (`npm ERR! 409 Conflict - GET https://registry.npmjs.org/astral-regex
`) seems unrelated as I'm not touching any dependencies: https://github.com/weiran-zsd/eslint-plugin-node/runs/6145701620?check_suite_focus=true#step:4:9 🤔

**Comment by aladdin-add** [comment]
no worries, I've re-run the ci - seems working now! 🎉

File: lib/rules/exports-style.js [hunk 1]
```diff
@@ -139,6 +139,87 @@ function getExportsNodes(scope) {
...
+        node.parent.type === "MemberExpression" && node.parent.object === node
+    )
+}
+
+function fixModuleExports(node, sourceCode, fixer) {
+    if (isModuleExportsReference(node)) {
+        return fixer.replaceText(node, "exports")
+    }
+    if (!isModuleExportsObjectAssignment(node)) {
+        return undefined
```
Comment (aladdin-add) [inline lib/rules/exports-style.js:207]:
```suggestion
        return null
```

File: lib/rules/exports-style.js [hunk 2]
```diff
@@ -139,6 +139,87 @@ function getExportsNodes(scope) {
...
+    }
+    const statements = []
+    const properties = node.parent.right.properties
+    for (const property of properties) {
+        const statement = getReplacementForProperty(property, sourceCode)
+        if (statement) {
+            statements.push(statement)
+        } else {
+            // No replacement available, give up on the whole thing
+            return undefined
```
Comment (aladdin-add) [inline lib/rules/exports-style.js:217]:
```suggestion
            return null
```

File: tests/lib/rules/exports-style.js [hunk 3]
```diff
@@ -146,14 +156,154 @@ new RuleTester().run("exports-style", rule, {
...
+            code: "a = module.exports.foo + module.exports['bar']",
+            output: "a = exports.foo + exports['bar']",
+            options: ["exports"],
+            parserOptions: { ecmaVersion: 8 },
+            globals: { module: false, exports: true },
+            errors: [
+                "Unexpected access to 'module.exports'. Use 'exports' instead.",
+                "Unexpected access to 'module.exports'. Use 'exports' instead.",
+            ],
+        },
```
Comment (aladdin-add) [inline tests/lib/rules/exports-style.js:303]: can you add a few more cases for non-top-level exports, e.g.

```js
if(foo){ module.exports = { foo: 1};}
else { module.exports = {foo: 2};}
```

**Review by aladdin-add — CHANGES_REQUESTED** [review]
just a few minor suggestions, thx for the contributing! ~

File: lib/rules/exports-style.js [hunk 4]
```diff
@@ -139,6 +139,87 @@ function getExportsNodes(scope) {
...
 }
 
+function getReplacementForProperty(property, sourceCode) {
+    if (property.kind === "get" || property.type === "SpreadElement") {
+        // module.exports = { get foo() { ... } }
+        // module.exports = { foo }
+        // module.exports = { ...foo }
+        // We don't have a similarly nice syntax for adding these directly
+        // on the exports object. Give up on fixing the whole thing.
+        return undefined
```
Comment (aladdin-add) [inline lib/rules/exports-style.js:149]: it's more common to return null in fixers.

```suggestion
        return null
```

File: lib/rules/exports-style.js [hunk 5]
```diff
@@ -139,6 +139,87 @@ function getExportsNodes(scope) {
     return variable.references.map(reference => reference.identifier)
 }
 
+function getReplacementForProperty(property, sourceCode) {
+    if (property.kind === "get" || property.type === "SpreadElement") {
+        // module.exports = { get foo() { ... } }
```
[the remark this answers is not in the document]
Comment (papandreou) [reply lib/rules/exports-style.js]: Good idea to make it more defensive like that! Changed in e783167d7f2496d24209fcb2f9cacabb451659e7

We do have good ways to deal with shorthands and computeds, though, see:

* https://github.com/papandreou/eslint-plugin-node/blob/8425c8063e3310b19c5c7eeb20f44c6b185ee7cd/tests/lib/rules/exports-style.js#L244
* https://github.com/papandreou/eslint-plugin-node/blob/8425c8063e3310b19c5c7eeb20f44c6b185ee7cd/tests/lib/rules/exports-style.js#L264

File: lib/rules/exports-style.js (hunk 1 above)
Comment (papandreou) [reply lib/rules/exports-style.js]: a4ec5000a715ec5c698e487bb685a9b2987b5a1f

File: lib/rules/exports-style.js (hunk 2 above)
Comment (papandreou) [reply lib/rules/exports-style.js]: a4ec5000a715ec5c698e487bb685a9b2987b5a1f

File: lib/rules/exports-style.js (hunk 4 above)
Comment (papandreou) [reply lib/rules/exports-style.js]: a4ec5000a715ec5c698e487bb685a9b2987b5a1f

File: tests/lib/rules/exports-style.js (hunk 3 above)
Comment (papandreou) [reply tests/lib/rules/exports-style.js:332]: Good shout, 8425c8063e3310b19c5c7eeb20f44c6b185ee7cd

**Review by aladdin-add — APPROVED** [review]
LGTM, thanks!

**Comment by aladdin-add** [comment]
released in v15.2.0

**Pull request merged**